src/main/kotlin/sliv/tool/scene/view/SceneView.kt

```diff
@@ -28,22 +35,59 @@ class SceneView : View() {
         //TODO: get size from the first image after data virtualization will be done
         val width = 500.0
         val height = 500.0
-
         val margin = 10.0
 
+        val scaleProperty = SimpleDoubleProperty(1.0)
+
         val columnsNumber = 30 //TODO: should be set from the UI
 
         val grid = VirtualizedFXGridProvider.createGrid(
-            frames,
-            columnsNumber,
-            width + margin,
-            height + margin
+            frames, columnsNumber, width + margin, height + margin, scaleProperty
         ) { frame ->
             FrameView(
-                width, height, frame
+                width, height, scaleProperty, frame
             )
         }
 
-        add(grid)
+        grid.setUpPanning()
+
+        grid.setOnScroll { event ->
+            if (event.isConsumed) { // If event is consumed by vsp
+                return@setOnScroll
+            }
+            if(event.deltaY.compareTo(0) == 0) {
+                return@setOnScroll
+            }
+            zoomGrid(scaleProperty, grid, event.x to event.y, event.deltaY > 0)
+        }
+
+        add(grid.node)
+    }
+
+    private fun zoomGrid(scaleProperty: DoubleProperty, grid: Grid, mousePos: Pair<Double, Double>, isPositive: Boolean) {
+        val initialPos = grid.getPosition()
+
+        val initialMouseX = (initialPos.first + mousePos.first) / scaleProperty.value
+        val initialMouseY = (initialPos.second + mousePos.second) / scaleProperty.value
+
+        if (isPositive) {
```

Comment:
why we need this? can we just check both borders instead?

Reply:
The body of this if statement not only checks the border, it also changes value for the scale property, so this method should know if scale should grow or decrease